### PR TITLE
WhitelistResolver must be invoked before other import resolvers

### LINKED ISSUE #1707: [BUG]: Linting raises library-not-found problem when import `type_extensions`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

We raise a library-not-found problem when import type_extentions

```
if TYPE_CHECKING:
    from typing_extensions import ParamSpec

    P = ParamSpec("P")
```

### Expected Behavior

No problem should be raised. Maybe we should white list this library

### Steps To Reproduce

1. Create unit test for above code snippet

### Cloud

Azure

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Ensure that WhitelistResolver, if present, is always imported first

### Linked issues
Resolves #1707 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Review by pritishpai — APPROVED** [review]
LGTM!

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
Could we revert the list? Now "first" is the last resolver in the list, which reads confusing

File: src/databricks/labs/ucx/source_code/graph.py [hunk 1]
```diff
@@ -369,10 +369,17 @@ def _chain_library_resolvers(library_resolvers: list[BaseLibraryResolver]) -> Ba
 
     @staticmethod
     def _chain_import_resolvers(import_resolvers: list[BaseImportResolver]) -> BaseImportResolver:
+        # check that WhiteListResolver comes first when resolving import
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/graph.py:372]: we currently have:
image

```mermaid
graph TD

PathLookup -->|inject into| resolve_import
DependencyResolver -->|call|resolve_import

resolve_import -.->|method of| BaseImportResolver
BaseImportResolver -.->|subclass| LocalFileResolver
BaseImportResolver -.->|subclass| WhitelistResolver
BaseImportResolver -.->|subclass| StubImportResolver
LocalFileResolver -.->|next| WhitelistResolver
WhitelistResolver -.->|next| StubImportResolver

Whitelist --> WhitelistResolver
FileLoader --> LocalFileResolver
```
this is hard to work with and requires code like this to fix flaw on the higher level.

# Simplify

```mermaid
graph TD

PathLookup -->|inject into| resolve_import
DependencyResolver -->|call|resolve_import
resolve_import -.->|method of| ImportResolver
Whitelist -->|inject into constructor| ImportResolver
FileLoader --> |inject into constructor| ImportResolver
```

**Review by nfx — CHANGES_REQUESTED** [review]
instead of making a hack, make the structure easy to work with.

```mermaid
graph TD

PathLookup -->|inject into| resolve_import
DependencyResolver -->|call|resolve_import
resolve_import -.->|method of| ImportResolver
Whitelist -->|inject into constructor| ImportResolver
FileLoader --> |inject into constructor| ImportResolver
```

**Comment by ericvergnaud** [comment]
superseded by #1746

**Pull request closed without merging**